File: ModelIO_Bindings/ModelIO_Bindings/MDLVertexDescriptor.swift
//
//  MDLVertexDescriptor.swift
//  ModelIO_Bindings
//
//  Created by Taanish Reja on 1/8/25.
//

import MetalKit
import ModelIO

// MDL Vertex Attribute
// create
public func createMDLVertexAttribute(namePtr: UnsafePointer<CChar>, rawFormat: UInt, offset: Int, bufferIndex: Int) -> UnsafeMutableRawPointer
{
    let name = String(cString: namePtr);
    let format = MDLVertexFormat(rawValue: rawFormat)!;
    let mdlVertexAttribute = MDLVertexAttribute(name: name, format: format, offset: offset, bufferIndex: bufferIndex);
    let mdlVertexAttributePtr = UnsafeMutableRawPointer(Unmanaged.passRetained(mdlVertexAttribute).toOpaque());
    return mdlVertexAttributePtr;
}

// release
public func releaseMDLVertexAttribute(mdlVertexAttributePtr: UnsafeMutableRawPointer)
{
    let mdlVertexAttribute = Unmanaged<MDLVertexAttribute>.fromOpaque(mdlVertexAttributePtr);
    mdlVertexAttribute.release();
}

// MDL Vertex Descriptor
// create
public func createMDLVertexDescriptor() -> UnsafeMutableRawPointer
{
    let mdlVertexDescriptor = MDLVertexDescriptor();
    let mdlVertexDescriptorPtr = Unmanaged<MDLVertexDescriptor>.passRetained(mdlVertexDescriptor).toOpaque();
    return mdlVertexDescriptorPtr;
}

// MDL Vertex Layout
// set layout
public func setMDLVertexLayout(mdlVertexDescriptorPtr: UnsafeMutableRawPointer, index: Int, stride: Int)
{
    let mdlVertexDescriptor = Unmanaged<MDLVertexDescriptor>.fromOpaque(mdlVertexDescriptorPtr).takeUnretainedValue();
    let newLayout = MDLVertexBufferLayout(stride: stride);
    mdlVertexDescriptor.layouts[index] = newLayout;
}

// set vertex attribute
public func setMDLVertexDescriptorAttribute(mdlVertexDescriptorPtr: UnsafeMutableRawPointer, index: Int, mdlVertexAttributePtr: UnsafeMutableRawPointer)
{
    let mdlVertexDescriptor = Unmanaged<MDLVertexDescriptor>.fromOpaque(mdlVertexDescriptorPtr).takeUnretainedValue();
    let mdlVertexAttribute = Unmanaged<MDLVertexAttribute>.fromOpaque(mdlVertexAttributePtr).takeUnretainedValue(); // must release attribute seperately
    mdlVertexDescriptor.attributes[index] = mdlVertexAttribute;
}

// release
public func releaseMDLVertexDescriptor(mdlVertexDescriptor: UnsafeMutableRawPointer)
{
    let mdlVertexDescriptorPtr = Unmanaged<MDLVertexDescriptor>.fromOpaque(mdlVertexDescriptor);
    mdlVertexDescriptorPtr.release();
}

File: ModelIO_Bindings/ModelIO_Bindings/MDLVertexDescriptor.cpp
//
//  MDLVertexDescriptor.cpp
//  ModelIO_Bindings
//
//  Created by Taanish Reja on 1/9/25.
//

#include "MDLVertexDescriptor.h"

namespace ModelIO {
    MDLVertexLayout::MDLVertexLayout(int stride):
        m_stride{0}
    {}
        
    int MDLVertexLayout::getStride() const
    {
        return m_stride;
    }

    MDLVertexAttribute::MDLVertexAttribute(std::string name, MDLVertexFormat vertexFormat, int offset, int bufferIndex):
        m_name{name},
        m_vertexFormat{vertexFormat},
        m_offset{offset},
        m_bufferIndex{bufferIndex},
        m_vertexAttribute{
            std::make_shared<SwiftObject>(
            ModelIO_Bindings::createMDLVertexAttribute(name.c_str(), static_cast<unsigned long>(vertexFormat), offset, bufferIndex),
            ModelIO_Bindings::releaseMDLVertexDescriptor
            )
        }
    {}

    MDLVertexAttribute::~MDLVertexAttribute() = default;
    
    const std::string& MDLVertexAttribute::getName() const
    {
        return m_name;
    }

    MDLVertexFormat MDLVertexAttribute::getVertexFormat() const
    {
        return m_vertexFormat;
    }

    int MDLVertexAttribute::getOffset() const
    {
        return m_offset;
    }

    int MDLVertexAttribute::getBufferIndex() const
    {
        return m_bufferIndex;
    }

    SwiftObject* MDLVertexAttribute::getVertexAttribute()
    {
        return m_vertexAttribute.get();
    }

    MDLVertexDescriptor::MDLVertexDescriptor():
        m_vertexDescriptor{
            std::make_shared<SwiftObject>(
                 ModelIO_Bindings::createMDLVertexDescriptor(),
                 ModelIO_Bindings::releaseMDLVertexDescriptor
            )
        }
    {}

    MDLVertexDescriptor::~MDLVertexDescriptor() = default;

    void MDLVertexDescriptor::setAttribute(int index, MDLVertexAttribute& vertexAttribute)
    {
        if (index >= m_attributes.size()) {
            m_attributes.push_back(vertexAttribute);
        }else {
            m_attributes[index] = vertexAttribute;
        }
        
        ModelIO_Bindings::setMDLVertexDescriptorAttribute(m_vertexDescriptor->get(), index, vertexAttribute.getVertexAttribute()->get());
    }

    const MDLVertexAttribute& MDLVertexDescriptor::getAttribute(int index) const
    {
        return m_attributes[index];
    }

    void MDLVertexDescriptor::setLayout(int index, MDLVertexLayout& vertexLayout)
{
        if (index >= m_layouts.size()) {
            m_layouts.push_back(vertexLayout);
        }else {
            m_layouts[index] = vertexLayout;
        }
        
        ModelIO_Bindings::setMDLVertexLayout(m_vertexDescriptor->get(), index, vertexLayout.getStride());
    }

    const MDLVertexLayout& MDLVertexDescriptor::getLayout(int index) const
    {
        return m_layouts[index];
    }

    SwiftObject* MDLVertexDescriptor::getVertexDescriptor()
    {
        return m_vertexDescriptor.get();
    }
}
